lib/rpc/src/tcp_client.rs

```diff
@@ -0,0 +1,463 @@
+// Copyright © 2021 University of Colorado. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use abomonation::{decode, encode};
+use alloc::borrow::ToOwned;
+use alloc::string::String;
+use alloc::{vec, vec::Vec};
+use log::{debug, trace, warn};
+
+use smoltcp::iface::EthernetInterface;
+use smoltcp::socket::{SocketHandle, SocketSet, TcpSocket, TcpSocketBuffer};
+use smoltcp::time::Instant;
+use smoltcp::wire::IpAddress;
+
+use kpi::io::FileInfo;
+use vmxnet3::smoltcp::DevQueuePhy;
+
+use crate::cluster_api::{ClusterClientAPI, ClusterError, NodeId};
+use crate::rpc::*;
+use crate::rpc_api::RPCClientAPI;
+
+const RX_BUF_LEN: usize = 4096;
+const TX_BUF_LEN: usize = 4096;
+
+pub struct TCPClient<'a> {
+    iface: EthernetInterface<'a, DevQueuePhy>,
+    sockets: SocketSet<'a>,
+    server_handle: Option<SocketHandle>,
+    server_ip: IpAddress,
+    server_port: u16,
+    client_port: u16,
+    client_id: NodeId,
+    req_id: u64,
+}
+
+impl TCPClient<'_> {
+    pub fn new<'a>(
+        server_ip: IpAddress,
+        server_port: u16,
+        iface: EthernetInterface<'a, DevQueuePhy>,
+    ) -> TCPClient<'a> {
+        TCPClient {
+            iface: iface,
+            sockets: SocketSet::new(vec![]),
+            server_handle: None,
+            server_ip: server_ip,
+            server_port: server_port,
+            client_port: 10110,
+            client_id: 0,
+            req_id: 0,
+        }
+    }
+}
+
+impl ClusterClientAPI for TCPClient<'_> {
+    /// Register with controller, analogous to LITE join_cluster()
+    /// TODO: add timeout?? with error returned if timeout occurs?
+    fn join_cluster(&mut self) -> Result<NodeId, ClusterError> {
+        // create client socket
+        let tcp_rx_buffer = TcpSocketBuffer::new(vec![0; RX_BUF_LEN]);
```

Comment:
We should try to handle OOM memory errors gracefully, we don't do a complete job at this yet in the kernel (which is my fault), but we should try for all future code... e.g. this will cause an implicit heap allocation, better would to use the try_with_capacity, try_extend, try_add etc. methods and return an error on oom